src/nifs/protogalaxy/mod.rs

```diff
@@ -42,18 +43,55 @@ pub(crate) struct Challenges<F: PrimeField> {
     pub gamma: F,
 }
 
+/// Wrap to properly (consistent with on-circuit) absorb inside ro
+struct PlonkInstanceWrapper<'l, C: CurveAffine>(&'l PlonkInstance<C>);
+
+impl<C: CurveAffine, D: PrimeField, RO: ROTrait<D>> AbsorbInRO<D, RO>
+    for PlonkInstanceWrapper<'_, C>
+{
+    fn absorb_into(&self, ro: &mut RO) {
+        let PlonkInstance {
+            W_commitments,
+            instances,
+            challenges,
+        } = &self.0;
+
+        ro.absorb_field_iter(
+            W_commitments
+                .iter()
+                .flat_map(|W_commitment| {
+                    let BigUintPoint { x, y } = BigUintPoint::new(W_commitment).unwrap();
+                    x.into_iter().chain(y)
+                })
+                .chain(
+                    instances
+                        .iter()
+                        .flat_map(|instance| instance.iter())
+                        .copied(),
+                )
+                .chain(challenges.iter().copied())
+                .map(|b| crate::util::fe_to_fe(&b).unwrap()),
+        );
+    }
+}
+
 impl<F: PrimeField> Challenges<F> {
     #[instrument(skip_all)]
-    pub(crate) fn generate_one<'i, D: PrimeField, RO: ROTrait<D>, C: CurveAffine>(
+    fn generate_one<'i, D: PrimeField, RO: ROTrait<D>, C: CurveAffine>(
```

Comment:
NBD: this name might be confused with generate output=1 instead generate one random number